### PR TITLE
feat: implement mini-batch & continue-as-new for replicator workflow

### PR DESCRIPTION
**Pull request by bestmike007**
### What changed? Why?

+ Implement mini batches, which can be distributed to different workers
+ Continue workflow as new when checkpoint size reached, in order to prevent workflow timeout
 
### How did you test the change?

+ GitHub Actions
+ `go run ./cmd/admin workflow start --workflow replicator --input '{"StartHeight": 1, "EndHeight": 10001}' --blockchain ethereum --network mainnet --env local`

### CODE REVIEW

File: internal/workflow/replicator.go [hunk 1]
```diff
@@ -106,26 +121,102 @@ func (w *Replicator) execute(ctx workflow.Context, request *ReplicatorRequest) e
...
+					EndHeight:   batchEnd,
+					Parallelism: parallelism,
+					Compression: dataCompression,
+				})
+				if err != nil {
+					return xerrors.Errorf("failed to replicate block from %d to %d: %w", batchStart, batchEnd, err)
+				}
+			}
+
+			// Phase 3: update watermark
```
Comment (HenryYYang) [inline internal/workflow/replicator.go:207]: I think we need to validate the output to ensure the blocks are contiguous before we update the watermark.  If we use the metadata from the downloaded blocks, then we don't need to call getBlocksByHeight during the update watermark, which would further reduce the chance of error in the corner case of a reorg.

Comment (bestmike007) [reply internal/workflow/replicator.go:207]: Added validation before updating watermark.

File: internal/workflow/activity/update_watermark.go [hunk 2]
```diff
@@ -30,8 +31,9 @@ type (
 	}
 
 	UpdateWatermarkRequest struct {
-		Tag         uint32
-		BlockHeight uint64
+		Tag           uint32
+		BlockHeight   uint64
+		ValidateSince uint64
```
Comment (HenryYYang) [inline internal/workflow/activity/update_watermark.go:36]: nit: maybe validateStart would be a better name.  Please also add a comment on whether this is inclusive or exclusive index.

File: internal/workflow/activity/update_watermark.go [hunk 3]
```diff
@@ -63,13 +65,30 @@ func (a *UpdateWatermark) execute(ctx context.Context, request *UpdateWatermarkR
...
+	if validateStart <= 0 {
+		validateStart = 1
+	}
+	blocks, err := a.metaStorage.GetBlocksByHeightRange(ctx, tag, validateStart, request.BlockHeight+1)
 	if err != nil {
-		return nil, xerrors.Errorf("failed to get block by tag %d height %d: %w", tag, request.BlockHeight, err)
+		return nil, xerrors.Errorf("failed to get blocks by tag %d: %w", tag, err)
+	}
+	if len(blocks) > 1 {
+		if err := parser.ValidateChain(blocks[1:], blocks[0]); err != nil {
```
Comment (HenryYYang) [inline internal/workflow/activity/update_watermark.go:87]: the ValidateChain order is reversed here.

File: internal/workflow/activity/update_watermark.go [hunk 4]
```diff
@@ -63,13 +65,30 @@ func (a *UpdateWatermark) execute(ctx context.Context, request *UpdateWatermarkR
...
+		if request.ValidateSince >= request.BlockHeight {
+			return nil, xerrors.Errorf("ValidateSince %d should be smaller than BlockHeight %d",
+				request.ValidateSince, request.BlockHeight)
+		}
+		validateStart = request.ValidateSince
+	}
+	if validateStart <= 0 {
+		validateStart = 1
+	}
+	blocks, err := a.metaStorage.GetBlocksByHeightRange(ctx, tag, validateStart, request.BlockHeight+1)
```
Comment (HenryYYang) [inline internal/workflow/activity/update_watermark.go:82]: The GetBlocksByHeightRange will sort the blocks, so trying to get the blocks in reverse order is not allowed, so it'll fail here.

File: internal/workflow/activity/update_watermark.go [hunk 5]
```diff
@@ -63,13 +65,30 @@ func (a *UpdateWatermark) execute(ctx context.Context, request *UpdateWatermarkR
...
-	block, err := a.metaStorage.GetBlockByHeight(ctx, tag, request.BlockHeight)
+	validateStart := request.BlockHeight - 1
+	if request.ValidateSince > 0 {
+		if request.ValidateSince >= request.BlockHeight {
+			return nil, xerrors.Errorf("ValidateSince %d should be smaller than BlockHeight %d",
+				request.ValidateSince, request.BlockHeight)
+		}
+		validateStart = request.ValidateSince
+	}
+	if validateStart <= 0 {
```
Comment (HenryYYang) [inline internal/workflow/activity/update_watermark.go:79]: nit, this is just else.

Comment (bestmike007) [reply internal/workflow/activity/update_watermark.go:79]: It's not, the previous if clause is to override the default value of validateStart.

**Review by HenryYYang — APPROVED** [review]

**Pull request merged**